platform/engineering/design-docs/2021-05-18-forms-library.md

```diff
@@ -0,0 +1,942 @@
+    schema
+- Paring `schema` and `uiSchema` makes sharing reusable patterns awkward
+  - Implementation details such as the data paths are baked into the
+    definitions of these reusable patterns
+
+### High Level Design
+Formulate will be a modular library consisting of helper functions and
+components to aid in the rapid development of forms on VA.gov. These functions
+and components can be conceptualized as **sub-modules** for:
+- [Building form pages](#form-page-builder)
+  - Wrappers for Formik components using the design system
+  - Re-exporting necessary Formik components
+- [Routing](#routing)
+  - Simple set-up for using Formik across multiple pages
+  - Components for rendering the standard form "chrome"
+  - Looping through a series of pages per item in an array
+  - Conditionally-rendered pages
+- [Save in progress](#save-in-progress)
+  - Manual save
+  - Auto save
+  - Data migrations
+
+**Each sub-module will be independent.** If a form doesn't need the multi-page
+routing, the team building it simply won't use the routing sub-module, and it
+won't be included in their application bundle. Likewise, if a form needs more
+complex routing, the team building it can opt to roll their own, but still reap
+the benefits of using the rest of the system.
+
+_Note: "sub-modules" here refer to independent parts of a larger system and have
+nothing to do with [Git
+Sub-Modules](https://git-scm.com/book/en/v2/Git-Tools-Sub-Modules)._
+
+Additionally, the library will provide a number of helper functions and
+components for:
+- Field-level schema validation
+- [Building a review page](#json-schema-validation)
+
+## Specifics
+
+### Detailed Design
+**Key technologies** we'll use:
+- React
+- TypeScript
+- [`api-extractor`](https://api-extractor.com/pages/overview/intro/)
+- [React router](https://reactrouter.com/)
+- [Formik](https://formik.org/)
+
+The following sections outline the various parts of the system.
+
+#### Form page builder
+Formulate will leverage Formik for much of the form page building. It will
+**re-export Formik components** where it makes sense and **provide light
+wrappers for Formik input fields** using design system components.
+
+In addition to simple wrappers, Formulate will use Formik to provide more
+complex fields such as file upload which may be imported and used in
+applications.
+
+##### Why not just use Formik directly?
+It should be possible to use Formik directly without any issues. The form page
+builder sub-module intends to make using the design system with Formik simpler.
+
+#### Routing
+This sub-module will be responsible for the user navigating through a multi-page
+form. In particular, it will:
+- Present the form broken down into chapters and pages
+- Manage routes for each page
+- Render each page wrapped in the chrome appropriate for navigation
+  - Form title, page title, progress bar, back / continue buttons, etc.
+- Prevent progress through the form if there are validation errors on the
```

Comment:
This is something that the form page builder module probably needs. Or at least, anyone using the form page builder on its own will need this functionality.

Reply:
This is specifically in the routing section because it prevents routing through the form. [Formik handles validation when the form is submitted](https://formik.org/docs/guides/validation#when-does-validation-run), so for the form page builder, my thought is applications would just use Formik's submission result to determine what it should do.